### ISSUE TITLE
"Host" in Docker configuration not read when connecting to Mariadb

### ISSUE THREAD

**Issue by francesco-santini** [GoteoFoundation/goteo#207]
With a goteo:live branch I am trying to install Goteo as Docker containers.
This is my configuration (settings.yml) concerning the db:

> db:
>    driver:   mysql     # Database driver (mysql)
>    host:     mariadb # Database host
>    port:     3306      # Database port
>    charset:  utf8mb4     # Database charset
>    database: goteo     # Database schema (database name)
>    username: goteo      # Database user for the goteo database
>    password: goteo      # Password for the goteo database

With such a configuration however I got a "Connection refused" error from PHP to MySQL.

Hence I manually debugged /src/Goteo/Core/DB.php, which is the file throwing such an exception when preparing the connection to the DB.
The piece of code is 

> 35                 $dsn = Config::get('dsn');
> 36                 $username = Config::get('db.username', true);
> 37                 $password = Config::get('db.password', true);
> 38           parent::__construct(
> 39                 $dsn,
> 40                 $username,
> 41                 $password,
> 42                 // Avoid stric mysql configurations (sorry we're not there yet, so many old queries)
> 43                 array(\PDO::MYSQL_ATTR_INIT_COMMAND => "SET @@SESSION.sql_mode=''")
> 44             ); 

By printing $dsn I discovered that it is equal to 

> mysql:host=127.0.0.1;dbname=goteo;port=3306;charset=utf8mb4

which means that "mariadb" is not taken from the settings.yml file (reported at top of this post), but 127.0.0.1 is taken instead (no idea from where it is taken).
In this way, the PHP docker container tries to connect to itself, and not to the Mariadb docker, where mysql is actually executed.

Notice that I logged to the PHP docker container and I successfully managed to connect to MySQL in the MariaDB container (172.19.0.2 is the IP address of the Docker container in the Docker network).

> mysql -h 172.19.0.2 -u goteo -p goteo

Any idea about how to open the connection to the right host? mariadb and not 127.0.0.1 ?

**Comment by francesco-santini** [comment]
[UPDATE]

In src/Goteo/Core/DB.php I manually hardcoded the string 

> "mysql:host=mariadb;dbname=goteo;port=3306;charset=utf8mb4"

so forcing the host to be "mariadb" instead of "127.0.0.1"

The code now is:

>35 $dsn = Config::get('dsn');
>36 $username = Config::get('db.username', true);
>37 $password = Config::get('db.password', true);
>38 parent::__construct(
>39 "mysql:host=mariadb;dbname=goteo;port=3306;charset=utf8mb4",
>40 $username,
>41 $password,
>42 // Avoid stric mysql configurations (sorry we're not there yet, so many old queries)
>43 array(\PDO::MYSQL_ATTR_INIT_COMMAND => "SET @@SESSION.sql_mode=''")
>44 );


Now I do not get "Connection refused", but I get 

>Goteo\Application\Config\ConfigException: Database misconfiguration: SQLSTATE[HY000] [1045] Access denied for user >'goteo'@'172.19.0.2' (using password: YES)
> at Goteo.Core.DB.__construct(DB.php:73)
> at Goteo.Core.Model.factory(Model.php:78)
> at Goteo.Application.Config.load(Config.php:75)
> at (main)(index_dev.php:32)

where 172.19.0.2 is the IP address of the PHP docker. However, I had granted rights to it directly in the mysql docker container:

> GRANT ALL PRIVILEGES ON goteo.* TO 'goteo'@'%';

So PHP still does not manage to connect to MySQL.


But I can do that from inside the PHP docker container to the MAriaDB docker container:

Checking the container ids
>ubuntu@crowdfunding:\~$ docker ps
>CONTAINER ID   IMAGE                    COMMAND                  CREATED        STATUS          PORTS                                                                                                                           NAMES
>274928cacc1b   goteo-live_php           "/usr/local/bin/entr…"   47 hours ago   Up 31 minutes   9000/tcp, 35729/tcp                                                                                                             goteo-php
>0d5e4d87f29f   mariadb:10.2             "docker-entrypoint.s…"   11 days ago    Up 31 minutes   0.0.0.0:3307->3306/tcp, :::3307->3306/tcp                                                                                       goteo-mariadb
>afd7f3c789fb   mailhog/mailhog:latest   "MailHog"                11 days ago    Up 31 minutes   0.0.0.0:1025->1025/tcp, :::1025->1025/tcp, 0.0.0.0:8082->8025/tcp, :::8082->8025/tcp                                            goteo-mailhog
>0bc3d911b713   nginx:alpine             "/docker-entrypoint.…"   11 days ago    Up 31 minutes   0.0.0.0:35729->35729/tcp, :::35729->35729/tcp, 0.0.0.0:8081->80/tcp, :::8081->80/tcp, 0.0.0.0:8443->443/tcp, :::8443->443/tcp   goteo-webserver


Finding the IP address of the MAriaDB docker container
>ubuntu@crowdfunding:~$ docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' 0d5e4d87f29f
>172.19.0.6

Getting a bash inside the PHP docker container
>ubuntu@crowdfunding:\~$ docker exec -it 274928cacc1b bash

Connecting to Mysql in the MariaDB container
>root@274928cacc1b:/application# mysql -h 172.19.0.6 -u goteo -p goteo
>Enter password: 
>Reading table information for completion of table and column names
>You can turn off this feature to get a quicker startup with -A
>
>Welcome to the MySQL monitor.  Commands end with ; or \g.
>Your MySQL connection id is 12
>Server version: 5.5.5-10.2.40-MariaDB-1:10.2.40+maria~bionic mariadb.org binary distribution
>
>Copyright (c) 2000, 2021, Oracle and/or its affiliates.
>
>Oracle is a registered trademark of Oracle Corporation and/or its affiliates. Other names may be trademarks of their respective owners.
>
>Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
>

Showing my username and grants on the goteo db
>mysql> show grants;
>+------------------------------------------------------------------------------------------------------+
>| Grants for goteo@%                                                                                   |
>+------------------------------------------------------------------------------------------------------+
>| GRANT USAGE ON *.* TO 'goteo'@'%' IDENTIFIED BY PASSWORD '*AFA50EBF056998E2672CC341C69E0C5A6F33EB3A' |
>| GRANT ALL PRIVILEGES ON `goteo`.* TO 'goteo'@'%'                                                     |
>+------------------------------------------------------------------------------------------------------+
>2 rows in set (0.00 sec)

**Comment by carlos-mg89** [comment]
Hi! What's your OS and OS version? What GIT branch are you working in?

**Comment by francesco-santini** [comment]
My os is Ubuntu 20.04.3 LTS (GNU/Linux 5.11.0-1016-oracle x86_64)
and the branch, as suggested in a previous opened issue is "live"

Thanks

**Comment by carlos-mg89** [comment]
You shouldn't have to be editing these kind of details by yourself when setting up the project for first time.

Would you please take a moment and describe the whole process that you've done until now?

- Software installed
- Commands executed
- Edited files (with the contents), like the config one
- And any other details you think might be useful

This branch is working and you shouldn't have to have these kind of troubles.

**Comment by francesco-santini** [comment]
Sure Carlos,
thanks

I cloned the live branch and created a "config/settings.yml", which I have also included in /etc/environment, which I report below

>PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin"
>GOTEO_CONFIG_FILE=config/settings.yml
>DEBUG=0 

The config/settings.yml is 
[settings.txt](https://github.com/GoteoFoundation/goteo/files/7100337/settings.txt)

I only touched the "secret", the "timezone", the "url" and finally the "db" properties as 

>db:
>driver: mysql # Database driver (mysql)
>host: mariadb # Database host
>port: 3306 # Database port
>charset: utf8mb4 # Database charset
>database: goteo # Database schema (database name)
>username: goteo # Database user for the goteo database
>password: goteo # Password for the goteo database


On the machine I only have the goteo-live folder, and Ubuntu is today-updated with "apt-get update" and "upgrade". The machine is dedicated to this (It is an ubuntu vmachine in the Oracle cloud), so I did not install anything else.

I installed docker and docker-compose by following the links here
https://goteofoundation.github.io/goteo/docs/developers/docker.html

After this configuration, I then executed "docker/up --build"

Clearly it said that the goteo database was empty so
I opened a bash shell in the mariadb container, and then I logged to mysql (-u root -p crowdfunding) where I manually created the goteo database with

> CREATE DATABASE goteo;

and added the rights for the user goteo:

> CREATE USER 'goteo'@'%' IDENTIFIED BY 'goteo';
> GRANT ALL PRIVILEGES ON goteo.* TO 'goteo'@'%';

I used the same values as in the settings.yml file.

Then I closed goteo (ctrl+c) and I restarted it again with "docker/up --build".
This time it filled the goteo db because the db was there (and empty), and it returned no error.

However, when I try to log to goteo I get 

>Goteo\Application\Config\ConfigException: Database misconfiguration: SQLSTATE[HY000] [2002] Connection refused
> at Goteo.Core.DB.__construct(DB.php:73)
> at Goteo.Core.Model.factory(Model.php:78)
> at Goteo.Application.Config.load(Config.php:75)
> at (main)(index_dev.php:32)

DB.php was the file I was trying to change, because it seems it tries to connect to 127.0.0.1 and not to the IP address of the mariadb container.
